### ISSUE TITLE
Support for Forward_Open vs. Large_Forward_Open requests

### ISSUE THREAD

**Issue by akgoins** [ros-drivers/odva_ethernetip#6]
The Session::createConnection() creates a Large_Forward_Open connection (0x5B) and uses the connection::createForwardOpenRequest() which creates a large forward open connection request which is a bit misleading.  Some devices do not support a large forward connection, so for the drivers to be generalizable there needs to be two functions in the Session class.  One to create a normal connection and another to create a large connection.  Or else, there needs to be a parameter passed in to specify which connection is made.

**Comment by pbeeson** [comment]
@akgoins 

Did you ever get this working with regular Forward Open.  Like replacing 0x5B with 0x54?  If so, I'd be extremely anxious to get your Yaskawa PLC code, as I'm banging my head to use this with a Yaskawa PCL too.

**Comment by akgoins** [comment]
Yes, I was able to get it running.  I had to do a lot of trouble shooting using wireshark to make sure that the data was being properly sent.  I will send you an example that should help.

**Comment by pbeeson** [comment]
That would be great.  I got basic reading working with OpENer, but there are large dropouts and I still haven't figured out writing.

Using this seems simpler, but so far it doesn't get past connecting as you had found.

**Comment by hcalitz** [comment]
I am trying to connect to a Yamaha HCU controller.  I have a similar issue with Large  Forward Open.  Any suggestions please? 

The error message I get is 

Error:   RR Data Command failed with status 1
at line 290 in .... /odva_ethernetip/src/session.cpp
terminate called after throwing an instance of 'std::runtime_error'

Wireshark shows:
Frame 48: 110 bytes on wire (880 bits), 110 bytes captured (880 bits) on interface 0
Ethernet II, Src: YamahaMo_00:03:7c (00:04:c6:00:03:7c), Dst: IntelCor_92:dd:52 (68:05:ca:92:dd:52)
Internet Protocol Version 4, Src: 192.168.0.1, Dst: 192.168.0.3
Transmission Control Protocol, Src Port: 44818, Dst Port: 56448, Seq: 29, Ack: 123, Len: 56
    Source Port: 44818
    Destination Port: 56448
    [Stream index: 3]
    [TCP Segment Len: 56]
    Sequence number: 29    (relative sequence number)
    [Next sequence number: 85    (relative sequence number)]
    Acknowledgment number: 123    (relative ack number)
    0101 .... = Header Length: 20 bytes (5)
    Flags: 0x018 (PSH, ACK)
    Window size value: 2798
    [Calculated window size: 2798]
    [Window size scaling factor: -2 (no window scaling used)]
    Checksum: 0xa1d5 [unverified]
    [Checksum Status: Unverified]
    Urgent pointer: 0
    [SEQ/ACK analysis]
    [Timestamps]
    TCP payload (56 bytes)
    [PDU Size: 56]
EtherNet/IP (Industrial Protocol), Session: 0x01056A80, Send RR Data
Common Industrial Protocol
    **Service: Unknown Service (0x5b) (Response)**
        1... .... = Request/Response: Response (0x1)
        .101 1011 = Service: Unknown (0x5b)
    Status: Connection failure: 
        General Status: Connection failure (0x01)
        Additional Status Size: 1 (words)
        Additional Status
    [Request Path Size: 2 (words)]
    [Request Path: Connection Manager, Instance: 0x01]
        [Path Segment: 0x20 (8-Bit Class Segment)]
            [001. .... = Path Segment Type: Logical Segment (1)]
            [...0 00.. = Logical Segment Type: Class ID (0)]
            [.... ..00 = Logical Segment Format: 8-bit Logical Segment (0)]
            [8-Bit Class Segment]
        [Path Segment: 0x24 (8-Bit Instance Segment)]
            [001. .... = Path Segment Type: Logical Segment (1)]
            [...0 01.. = Logical Segment Type: Instance ID (1)]
            [.... ..00 = Logical Segment Format: 8-bit Logical Segment (0)]
            [8-Bit Instance Segment]
CIP Connection Manager
    Service: Large Forward Open (Response)
        1... .... = Request/Response: Response (0x1)
        .101 1011 = Service: Large Forward Open (0x5b)
    Status: Connection failure, Extended: Invalid consuming application path
        General Status: Connection failure (0x01)
        Additional Status Size: 1 (words)
        Extended Status: Invalid consuming application path (0x012a)
        Additional Status
    Command Specific Data
        Connection Serial Number: 0x7deb
        Vendor ID: Unknown (0x1234)
        Originator Serial Number: 0x00006789
        Remaining Path Size: 0
        Reserved: 0x00

**Comment by hcalitz** [comment]
I am trying to connect to a Yamaha YHX  HCU controller.  When using Large_Forward_Open (0x5B) I get a connection error.  Using Forward_open (0x54) I get too much data error (0x15).   The controller manual specify:

Max input connection size: 1448 bytes with Large_Forward_Open, 509 bytes with Forward_Open
Max output connection size: 1448 bytes with Large_Forward_Open, 505 bytes with Forward_Open

So seems that both options are supported.  I am a bit lost :) 

I looked at the code, and got stuck here: 

static EIP_DWORD calcConnectionParams(EIP_UINT size, bool variable, EIP_BYTE priority,
    EIP_BYTE type, bool shared)

in forward_open_request.h  

Any suggestion how to move past this sticky point please?
